### PR TITLE
Make `MODULE.bazel.lock` deterministic

### PR DESCRIPTION
**Pull request by fmeum**
Sorts the extensions, which would otherwise be added in the non-deterministic order in which module extension resolution events are fired, by their ID.

Since Gson doesn't support custom comparators on a `TreeMap`, `ModuleExtensionId` now implements `Comparable` directly.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java [hunk 1]
```diff
@@ -123,7 +124,7 @@ private ImmutableMap<ModuleExtensionId, LockFileModuleExtension> combineModuleEx
     for (ModuleExtensionResolutionEvent extensionEvent : extensionResolutionEvents) {
       updatedExtensionMap.put(extensionEvent.getExtensionId(), extensionEvent.getModuleExtension());
     }
-    return updatedExtensionMap.buildKeepingLast();
+    return ImmutableSortedMap.copyOf(updatedExtensionMap.buildKeepingLast());
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java:127]: I believe this entire PR can be reduced to this one line. (well, and using the lexicographical comparator explicitly.) ImmutableMap already has a stable iteration order so we don't need to hold on to it as a sorted map (especially as the iteration order is only ever important for serialization).

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java]: Well, that's much nicer now, thanks!

**Review by meteorcloudy — APPROVED** [review]
Nit: is there any simple way to test this?

**Comment by fmeum** [comment]
> Nit: is there any simple way to test this?

While not a dedicated test for this purpose, https://github.com/bazelbuild/bazel/pull/19343 generates a lockfile twice and compares the outcomes. We could have an additional test with more extensions, but it may still not fail reliably in the face of actual non-determinism.

**Comment by meteorcloudy** [comment]
OK, I'm fine with the test in https://github.com/bazelbuild/bazel/pull/19343

**Review by SalmaSamy — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by fmeum** [comment]
@SalmaSamy Just saw that Git authorship wasn't preserved during the commit. Not a big deal, but maybe the way this was merged internally needs to be adjusted for future commits?

**Comment by meteorcloudy** [comment]
Yes, we should not push the update mater branch by merge button for normal PRs.

**Comment by iancha1992** [comment]
@bazel-io fork 6.4.0

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 6.4.0 RC1](https://releases.bazel.build/6.4.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=last_rc.
Thanks!

**Pull request closed without merging**